aleph-client/src/contract/mod.rs

```diff
@@ -90,93 +79,109 @@ impl ContractInstance {
         &self.address
     }
 
-    /// The metadata of this contract instance.
-    pub fn ink_project(&self) -> &InkProject {
-        &self.ink_project
-    }
-
     /// Reads the value of a read-only, 0-argument call via RPC.
-    pub async fn contract_read0<T: Decode>(&self, conn: &Connection, message: &str) -> Result<T> {
-        self.contract_read(conn, message, &[]).await
+    pub async fn contract_read0<T: TryFrom<ConvertibleValue, Error = anyhow::Error>>(
+        &self,
+        conn: &Connection,
+        message: &str,
+    ) -> Result<T> {
+        self.contract_read::<String, T>(conn, message, &[]).await
     }
 
     /// Reads the value of a read-only call via RPC.
-    pub async fn contract_read<T: Decode>(
+    pub async fn contract_read<
+        S: AsRef<str> + Debug,
+        T: TryFrom<ConvertibleValue, Error = anyhow::Error>,
+    >(
         &self,
         conn: &Connection,
         message: &str,
-        args: &[&str],
+        args: &[S],
     ) -> Result<T> {
         let payload = self.encode(message, args)?;
         let args = ContractCallArgs {
             origin: self.address.clone(),
             dest: self.address.clone(),
             value: 0,
-            gas_limit: Weight {
-                ref_time: Self::MAX_READ_GAS,
-                proof_size: u64::MAX,
-            },
+            gas_limit: None,
             input_data: payload,
             storage_deposit_limit: None,
         };
-        conn.call_and_get(args)
+
+        let result = conn
+            .call_and_get(args)
             .await
-            .context("RPC request error - there may be more info in node logs.")
+            .context("RPC request error - there may be more info in node logs.")?
+            .result
+            .map_err(|e| anyhow!("Contract exec failed {:?}", e))?;
+        let decoded = self.decode(message, result.data)?;
+        ConvertibleValue(decoded).try_into()?
     }
 
     /// Executes a 0-argument contract call.
     pub async fn contract_exec0(&self, conn: &SignedConnection, message: &str) -> Result<()> {
-        self.contract_exec(conn, message, &[]).await
+        self.contract_exec::<String>(conn, message, &[]).await
     }
 
     /// Executes a contract call.
-    pub async fn contract_exec(
+    pub async fn contract_exec<S: AsRef<str> + Debug>(
         &self,
         conn: &SignedConnection,
         message: &str,
-        args: &[&str],
+        args: &[S],
+    ) -> Result<()> {
+        self.contract_exec_value(conn, message, args, 0).await
+    }
+
+    /// Executes a 0-argument contract call sending the given amount of value with it.
+    pub async fn contract_exec_value0(
+        &self,
+        conn: &SignedConnection,
+        message: &str,
+        value: u128,
```

Comment:
Is this in pico AZERO or AZERO? 

Also, nit: can we rename it to `amount` or `pazero_amount`?

Longer-term: How about we introduce the `AZERO` struct with proper constructors and accessors.

Reply:
> Longer-term: How about we introduce the AZERO struct with proper constructors and accessors.
 
All for this, it's very inconvenient to express values right now.

> Also, nit: can we rename it to amount or pazero_amount?

I don't like it. `Pico` and especially `p` isn't better for me personally, and we are just using these raw `u128` everywhere, so it wouldn't even be consistent.

Reply:
OK, so please add as part of the comment the unit.

Reply:
> OK, so please add as part of the comment the unit.

Same as with the name - I don't see the point. I'm not even sure about the unit myself (other than `1` here being the smallest value expressible as a `Balance`). It's not that I wouldn't like this to be expressed better  it's that I think having some mentions of units in this file, but not in others would make it worse. For example, currently grepping `aleph-client` by things like `pico`, `unit`, `decimal`, gives nothing.

I'd propose raising this as a separate ticket - "Unit Overhaul".